pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1331,57 +1364,62 @@ impl Timeline {
 
     /// Get cached size of relation if it not updated after specified LSN
     pub fn get_cached_rel_size(&self, tag: &RelTag, lsn: Lsn) -> Option<BlockNumber> {
-        let rel_size_cache = self.rel_size_cache.read().unwrap();
-        if let Some((cached_lsn, nblocks)) = rel_size_cache.map.get(tag) {
+        let rel_size_cache = self.rel_size_latest_cache.read().unwrap();
+        if let Some((cached_lsn, nblocks)) = rel_size_cache.get(tag) {
             if lsn >= *cached_lsn {
-                RELSIZE_CACHE_HITS.inc();
+                RELSIZE_LATEST_CACHE_HITS.inc();
                 return Some(*nblocks);
             }
             RELSIZE_CACHE_MISSES_OLD.inc();
         }
+        let mut rel_size_cache = self.rel_size_snapshot_cache.lock().unwrap();
+        if let Some(nblock) = rel_size_cache.get(&(lsn, *tag)) {
+            RELSIZE_SNAPSHOT_CACHE_HITS.inc();
+            return Some(*nblock);
+        }
         RELSIZE_CACHE_MISSES.inc();
```

Comment:
Let's add a label for this indicating whether the request was a primary or a replica.
It allows us to see if the new cache is effective.

Reply:
Done